File: app/src/main/java/me/eagzzycsl/darkroom/access/AppList.java
package me.eagzzycsl.darkroom.access;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import me.eagzzycsl.darkroom.db.SQLMan;
import me.eagzzycsl.darkroom.model.MetaApp;
import me.eagzzycsl.darkroom.model.NaughtyApp;
import me.eagzzycsl.darkroom.model.UserApp;


public final class AppList {
    private final static ArrayList<NaughtyApp> naughtyApps = new ArrayList<>();
    private final static ArrayList<UserApp> userApps = new ArrayList<>();
    private final static HashMap<String, MetaApp> apps = new HashMap<>();

    public static void init(Context context) {
        getAllUserApp(context);
        naughtyApps.clear();
        naughtyApps.addAll(SQLMan.getInstance(context).readAll());

    }

    public static MetaApp getMetaApp(String pkgName) {
        return apps.get(pkgName);
    }

    public static NaughtyApp getNaughtApp(String pkgName) {
        final MetaApp metaApp = getMetaApp(pkgName);
        return metaApp == null ? null : new NaughtyApp(metaApp);
    }

    public static void addMetaApp(MetaApp metaApp) {
        final String pkgName = metaApp.getPkgName();
        apps.putIfAbsent(pkgName, metaApp);
    }


    public static ArrayList<UserApp> getUserApps() {
        return userApps;
    }

    public static ArrayList<NaughtyApp> getNaughtyApps() {
        return naughtyApps;
    }

    public static void clearNaughtyApps() {
        naughtyApps.clear();
    }

    public static void addNaughtyApp(NaughtyApp naughtyApp) {
        naughtyApps.add(naughtyApp);
    }

    public static boolean useAppInNaughty(UserApp userApp) {
        return naughtyApps.stream().anyMatch(naughtyApp ->
                naughtyApp.getMetaApp() == userApp.getMetaApp()
        );
    }

    public static ArrayList<UserApp> getAllUserApp(Context context) {
        userApps.clear();
        PackageManager pm = context.getPackageManager();
        List<PackageInfo> packages = pm.getInstalledPackages(0);
        // 判断系统/非系统应用
        packages.stream()
                .filter(packageInfo -> (
                        packageInfo.applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) == 0
                ).
                forEach(packageInfo -> {
                    String pkgName = packageInfo.packageName;
                    MetaApp metaApp = AppList.getMetaApp(pkgName) == null ? new MetaApp(
                            packageInfo.applicationInfo.loadLabel(pm).toString(),
                            packageInfo.packageName,
                            packageInfo.applicationInfo.loadIcon(pm)
                    ) : AppList.getMetaApp(pkgName);
                    AppList.addMetaApp(metaApp);
                    userApps.add(new UserApp(metaApp));
                });
        return userApps;
    }
}

File: app/src/main/java/me/eagzzycsl/darkroom/manager/AppList.kt
package me.eagzzycsl.darkroom.manager

import android.content.Context
import android.content.pm.ApplicationInfo

import me.eagzzycsl.darkroom.db.SQLMan
import me.eagzzycsl.darkroom.model.MetaApp
import java.text.Collator
import java.util.*


object AppList {
    private val metaApps = HashMap<String, MetaApp>()
    val naughtyApps = ArrayList<MetaApp>()
    val sysApps = ArrayList<MetaApp>()
    val userApps = ArrayList<MetaApp>()
    val easyFreezeApps = ArrayList<MetaApp>()
    private val appNameCollator = Collator.getInstance(Locale.CHINA)
    fun init(context: Context) {
        // 读取数据库中app的是否冻结记录
        val appConfigRecords  = SQLMan.getAllAppConfig(context)
        // 获取手机上安装的全部app
        // 如果数据库中标记某个app为冻结，那么标记该app为冻结app
        val pm = context.packageManager
        val apps = pm.getInstalledApplications(0)
        apps.filter { it.flags and ApplicationInfo.FLAG_SYSTEM != 1 }.forEach({
            val metaApp = MetaApp(
                    it.loadLabel(pm).toString(),
                    it.packageName,
                    it.loadIcon(pm),
                    appConfigRecords[it.packageName]?.isNaughty?: false,
                    false
            )
            metaApps[it.packageName] = metaApp
        })
        apps.filter { it.flags and ApplicationInfo.FLAG_SYSTEM == 1 }.forEach({
            val metaApp = MetaApp(
                    it.loadLabel(pm).toString(),
                    it.packageName,
                    it.loadIcon(pm),
                    appConfigRecords[it.packageName]?.isNaughty?: false,
                    true
            )
            metaApps[it.packageName] = metaApp
        })
        this.calc()
        // 将新的数据写回数据库
        SQLMan.writeAllMetaApps(context, metaApps.values.toList())
    }
    private fun calc() {
        this.calcNaughtyApps()
        this.calcUserApps()
        this.calcSysApps()
    }
    private fun calcNaughtyApps() {
        this.naughtyApps.clear()
        this.naughtyApps.addAll(
                metaApps.filterValues { it.isNaughty  }.values.toList()
        )
        this.naughtyApps.sortWith(
                kotlin.Comparator { o1, o2 -> appNameCollator.compare(o1.appName, o2.appName) }
        )
    }
    private fun calcUserApps() {
        this.userApps.clear()
        this.userApps.addAll(
                metaApps.values.filter { !it.isSysApp }
        )
        this.userApps.sortWith(
                kotlin.Comparator { o1, o2 -> appNameCollator.compare(o1.appName, o2.appName) }
        )
    }
    private fun calcSysApps() {
        this.sysApps.clear()
        this.sysApps.addAll(
                metaApps.values.filter { it.isSysApp }
        )
        this.sysApps.sortWith(
                kotlin.Comparator { o1, o2 -> appNameCollator.compare(o1.appName, o2.appName) }
        )
    }
    fun genEasyFreezeApps(appName: String) {
        easyFreezeApps.clear()
        easyFreezeApps.addAll(
                metaApps.values.filter { it.appName == appName }
        )
    }

    fun saveNaughtyApps(context: Context) {
        SQLMan.markNaughtyApps(context,
                metaApps.values.filter { it.willFreeze }
        )
        this.calcNaughtyApps()
    }
    fun markFrozen() {
        this.metaApps.values.forEach { it -> it.willFreeze = it.isNaughty }
    }
    // TODO:这样子会导致已经冻结的app还没有被释放
    fun markNaughty() {
        this.metaApps.values.forEach { it-> it.isNaughty = it.willFreeze}
    }
    fun unMarkFrozen() {
        this.metaApps.values.forEach{ it -> it.willFreeze = false }
    }
}
